In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import warnings
warnings.filterwarnings("ignore")

<h3>Задача 1</h3>

Даны значения величины заработной платы заемщиков банка (salary) и значения их поведенческого кредитного скоринга (scoring):

salary = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110] \
scoring = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]
1. Найдите ковариацию этих двух величин: сначала без использования специальных функций, а затем с помощью функции numpy.cov. Полученные значения должны быть равны.
2. Найдите коэффициент корреляции Пирсона: с помощью ковариации и средних квадратических отклонений двух признаков, а затем с помощью специальных функций из библиотек numpy и pandas.

In [2]:
salary = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
scoring = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

*1) Найдем ковариацию этих двух величин: сначала без использования специальных функций. Для этого найдем средние значения*

In [3]:
salary_mean = salary.mean()
scoring_mean = scoring.mean()
print(f'salary_mean = {salary_mean}, scoring_mean = {scoring_mean}')

salary_mean = 101.4, scoring_mean = 709.9


In [4]:
cov = ((salary - salary_mean) * (scoring - scoring_mean)).sum() / (salary.shape[0] - 1)
cov

10175.37777777778

*Найдем ковариацию этих двух величин:  с помощью функции numpy.cov.*

In [5]:
np.cov(salary, scoring)

array([[ 3882.93333333, 10175.37777778],
       [10175.37777778, 33854.32222222]])

*Полученные значения совпадают.*

*2) Найдем коэффициент корреляции Пирсона: с помощью ковариации и средних квадратических отклонений двух признаков. Для начала нужно найти среднее квадратическое отклонение*

In [6]:
salary_std = np.sqrt(((salary - salary_mean)**2).sum() / salary.shape[0])
salary_std

59.115480206118605

In [7]:
scoring_std = np.sqrt(((scoring - scoring_mean)**2).sum() / scoring.shape[0])
scoring_std

174.55340157098058

*Найдем коэффициент корреляции Пирсона по формуле:*

$$r_{XY} = \frac{cov_{XY}}{\sigma_{X}\sigma_{Y}}$$

In [8]:
cov/(salary_std * scoring_std)

0.9861001023043514

*Получили результат при несмещенной дисперсии = 0,986. Посчитаем то же самое для смещенной дисперсии:*


In [25]:
salary_s = np.sqrt(((salary - salary_mean)**2).sum() / (salary.shape[0] - 1))
salary_s

62.31318747531162

In [27]:
scoring_s= np.sqrt(((scoring - scoring_mean)**2).sum() / (scoring.shape[0] - 1))
scoring_s

183.99544076477068

In [28]:
cov/(salary_s * scoring_s)

0.8874900920739164

*Найдем коэффициент корреляции Пирсона:с помощью специальных функций из библиотек numpy*

In [9]:
np.corrcoef(salary, scoring)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

*Найдем коэффициент корреляции Пирсона:с помощью специальных функций из библиотек pandas*

In [10]:
df = pd.DataFrame({'salary': [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], 
'scoring': [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]})
df.head()

salary  scoring
0      35      401
1      45      574
2     190      874
3     200      919
4      40      459

In [14]:
cov / (salary.std(ddof=1) * scoring.std(ddof=1))

0.8874900920739164

In [11]:
corr = df[['salary', 'scoring']].corr()
corr 

salary  scoring
salary   1.00000  0.88749
scoring  0.88749  1.00000

**Ответ:**
1. ковариация двух величин ('salary' и 'scoring') = 10175.37. Полученные значения разными путями равны. \
2. коэффициент корреляции Пирсона = 0.887
По коэффициенту корреляции Пирсона можно отметить достаточно высокую взаимозависимость двух величин 'salary' и 'scoring'.

<h3>Задача 2</h3>

Измерены значения IQ выборки студентов, обучающихся в местных технических вузах:

131, 125, 115, 122, 131, 115, 107, 99, 125, 111

Известно, что в генеральной совокупности IQ распределен нормально. Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [34]:
sample = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
p = 0.95
alpha = 1 - p

*т.к. дисперсия нам не известна будем использовать t-статистику*

In [35]:
n = sample.shape[0]
# среднее значение выборки:
mean = sample.mean()
# среднеквадратическое отклонение при несмещенной оценке:
std = sample.std(ddof=1)

*Найдем квантили, чтобы определить критическую область, используя t-статистику*

In [36]:
t1 = stats.t.ppf(alpha / 2, df = n - 1)
t2 = stats.t.ppf(1 - alpha / 2, df = n - 1)
t1, t2

(-2.2621571627409915, 2.2621571627409915)

*Теперь найдем доверительный интервал:*

In [38]:
(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))

(110.55608365158724, 125.64391634841274)

**Ответ:** доверительный интервал от 110,55 до 125,64

<h3>Задача 3</h3>

Известно, что рост футболистов в сборной распределен нормально с известной дисперсией 25. На выборке объёма 27 выборочное среднее составило 174.2. Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [40]:
n = 27
sigma = 25
mean = 174.2
p = 0.95
alpha = 1 - p

*т.к. дисперсия нам известна, будем использовать z-статистику*

*Найдем квантили, чтобы определить критическую область*

In [41]:
t1 = stats.norm.ppf(alpha / 2)
t2 = stats.norm.ppf(1 - alpha / 2)
t1, t2

(-1.959963984540054, 1.959963984540054)

Критическая область от - бесконечности до -1,959 и от 1,959 до + бесконечности

In [42]:
(mean + t1 * sigma / np.sqrt(n), mean + t2 * sigma / np.sqrt(n))

(164.77011888269857, 183.6298811173014)

**Ответ:** доверительный интервал от 164.77 до 183.63

<h3>Задача 4</h3>

Выберите тему для проектной работы по курсу Теории вероятностей и математической статистики и напишите ее в комментарии к Практическому заданию.

Планирую рассмотреть данные по кредитам. Исследовать взаимосвязи заемщиков к просроченной задолженности, проверю гипотезу, что чем выше сумма, тем меньше просрочка